# Scripts and comand lines for the computation of genomic traits  and the phylogenetic signals of these traits

### by Sara Beier with support of Johannes Werner

This file contains unix shell command lines and R scripts to compute genomic traits from the genome data for the example genome 2593339298 (Mameliella alba DSM 26384, [JGI/IMG platform](https://img.jgi.doe.gov/)) as well as phylogenetic signals from a list of genomic traits in reference genomes as used in the manuscript Beier et al, in prep, doi: !still missing!

## Data accession

the JGI/IMG platform provides the tar-compressed file 2593339298.tar.gz

```bash
$ tar -xzvf 2593339298.tar.gz

$ ls 2593339298
2593339298.cog.tab.txt  2593339298.genes.fna       2593339298.ipr.tab.txt   2593339298.signalp.tab.txt  README.txt
2593339298.fna          2593339298.gff             2593339298.ko.tab.txt    2593339298.tigrfam.tab.txt
2593339298.genes.faa    2593339298.intergenic.fna  2593339298.pfam.tab.txt  2593339298.tmhmm.tab.txt
```

Some genomic traits, such as the genome size, the number of genes, the number of 16s rRNA gene copies (RRN), the fraction of horizontally transferred genes (%HGT) are provided via the JGI genome statistic, while others (see code below) were computed from the genome sequence data.

## Transcription factors

This code returns total number of transcription factors. To obtain fraction of transcription factors (column %TF in Table S1) divide by the number of total genes (column 'Gene count' in Table S1) given via the JGI platform or in Table S1.
Input file: DBD-PFAM.list, list pfams with DNA binding sites from the DBD database (Wilson et al., 2008).

```bash
$ head DBD-PFAM.list 
#http://www.transcriptionfactor.org/Download/pfam_18.dbds.v2.03.txt
# generated by ./pfamdbddetails.pl 18 pfam_catfile_18
pfam00010
pfam00046
pfam00096
pfam00105
pfam00126
pfam00157
pfam00165
pfam00170

$ TF_count=$(grep -f DBD-PFAM.list 2593339298/2593339298.pfam.tab.txt |wc -l) #count transcription factors
$ echo $TF_count
182
```

Citation:  
* Wilson, D., Charoensawan, V., Kummerfeld, S.K., and Teichmann, S.A. (2008) DBDtaxonomically broad transcription factor predictions: new content and functionality. Nucleic Acids Res 36: D88â€“D92.

## GC content

This code returns the genomic GC content in % (adapted from https://www.biostars.org/p/70167/)(see column '%GC' in Table S1).

```bash
$ GC=$(awk 'BEGIN {FS="";cg=0;t=0;} \
{if ($1 != ">") {for (i = 1; i <= NF; i++) {if ($i ~ /[ACTGactg]/) {t++;
} \
if ($i ~ /[CGcg]/) {cg++;}}}} \
END {print cg"\t"t"\t"(cg/t)*100;}' 2593339298/2593339298.fna |cut -f3)
$ echo $GC
65.2171
```

## Prophages

This code returns the total number of category 1 (sure) and category 2 (somewhat sure) prophages detected via the VirSorter software (Roux et al., 2015) (see column'Prophages' in Table S1).

```bash
#run virsorter
$ wrapper_phage_contigs_sorter_iPlant.pl -f 2593339298/2593339298.fna --db 1 --wdir virout.2593339298 --ncpu 4 --data-dir /
virsorter-data/

#extract prophage counts from VirSorter outputfile
$ var1=$(sed -n '/## 4/,$p' virout.2593339298/VIRSorter_global-phage-signal.csv | wc -l) #counts number of lines including and below line: ## 4 - Prophages - category 1 (sure)
$ var2=$(sed -n '/## 5/,$p' virout.2593339298/VIRSorter_global-phage-signal.csv | wc -l) #counts number of lines including and below line: ## 5 - Prophages - category 2 (somewhat sure)
$ var3=$(sed -n '/## 6/,$p' virout.2593339298/VIRSorter_global-phage-signal.csv | wc -l) #counts number of lines including and below line: ## 6 - Prophages - category 3 (not so sure)
$ prophages=$(echo "($var1-$var2-2) + ($var2-$var3-2)"|bc) #returns number of detected prophages from catagory 1 and category 2
$ echo $prophages
4
```

Citation:  
* Roux, S., Enault, F., Hurwitz, B.L., and Sullivan, M.B. (2015) VirSorter: mining viral signal from microbial genomic data. Peerj 3: e985.

## Codon Usage bias (Vieira-Silva)

This code returns the codon usage bias parameter F that was used in the Figures 3 and 4 for CUB, as well as the predicted generation time d.vs as detailed elsewhere (Vieira-Silva 2010)(see column 'Generation time (Vieira-Silva)' in Table S1). The parameter F is based on the earlier defined parameters &Delta;ENC' (Rocha 2004) and S (Sharp et al. 2005). Codon usage bias summary statistics were calculated in a shell script using the ENCprime software (https://github.com/jnovembre/ENCprime). Sequences coding for ribosomal proteins were extracted using the script extractFromFasta.pl (https://github.com/jonbra/NGS-Abel/blob/master/scripts/extractFromFasta.pl). Perl code available via the FAS center page (http://archive.sysbio.harvard.edu/CSB/resources/computational/scriptome/UNIX/) was used to reformat sequence files.
The ENCprime output together with extracted ribosomal gene coding sequences were used as an input for an downstream R script, to return the variables F and d.vs.

```bash
#select annotated genes (without stopp-codon)
$ cd 2593339298
$ awk '/^>/ {printf("\n%s\n",$0);next; } { printf("%s",$0);}  END {printf("\n");}' < 2593339298.genes.fna >enc.outfile.fas
$ awk '{if ( $1 != "") print $1}' <enc.outfile.fas> enc.tmp
$ grep '>' 2593339298.genes.faa |sed 's/>//'g |sed 's/\s.*$//' > enc.2593339298.genes.list
$ perl ../extractFromFasta.pl enc.tmp list enc.2593339298.genes.list  > enc.tmp2
$ sed 's/TAA$//' enc.tmp2 |sed 's/TGA$//'|sed 's/TAG$//'|sed 's/^ATG//' >enc.2593339298.genesnostop.fna #remove final stop-codons and initial start codons

#remove sequences which are not multiple of 3
$ perl -e '$count=0; $len=0; while(<>) {s/\r?\n//; s/\t/ /g; if (s/^>//) { if ($. != 1) {print "\n"} s/ |$/\t/; $count++; $_ .= "\t";} else {s/ //g; $len += length($_)} print $_;} print "\n"; warn "\nConverted $count FASTA records in $. lines to tabular format\nTotal sequence length: $len\n\n";' enc.2593339298.genesnostop.fna  > enc.c.2593339298.tab #transform into tab-file
$ perl -e ' $col=2; while (<>) { s/\r?\n//; @F = split /\t/, $_; $len = length($F[$col])/3; print "$_\t$len\n" } warn "\nAdded column with length of column $col for $. lines.\n\n"; ' enc.c.2593339298.tab|awk -F '\t' '$4 ~ /^[0-9]+$/ { print $0 }'  |cut -f1-3 >enc.tc.2593339298.tab #select sequences which are a multiple of 3
$ perl -e ' $len=0; while(<>) { s/\r?\n//; @F=split /\t/, $_; print ">$F[0]"; if (length($F[1])) { print " $F[1]" } print "\n"; $s=$F[2]; $len+= length($s); $s=~s/.{60}(?=.)/$&\n/g; print "$s\n"; } warn "\nConverted $. tab-delimited lines to FASTA format\nTotal sequence length: $len\n\n"; ' enc.tc.2593339298.tab |sed 's/\s.*$//'> enc.tc.2593339298.genes.fna #change back to fasta and remove first space and everything after from seqids

#select sequences coding for ribosomal proteins
$ grep 'Ribosomal protein' 2593339298.cog.tab.txt|cut -f 1 >enc.2593339298.rib.list
$ perl ../extractFromFasta.pl enc.tc.2593339298.genes.fna list enc.2593339298.rib.list  > enc.tc.2593339298.ribgenes.fna

#Concatenate multiple-sequence fasta file to single long sequence
$ grep -v "^>" enc.tc.2593339298.genes.fna | awk 'BEGIN { ORS=""; print ">con_all\n" } { print }' > enc.tc.2593339298.con-all.fna
$ sed -i -e '$a\' enc.tc.2593339298.con-all.fna
$ grep -v "^>" enc.tc.2593339298.ribgenes.fna | awk 'BEGIN { ORS=""; print ">con_rib\n" } { print }' > enc.tc.2593339298.con-rib.fna
$ sed -i -e '$a\' enc.tc.2593339298.con-rib.fna

#join concatenated sequence files and a file with non-rib seqs (all sequences are doubled --> important to keep constant ratio between sequences in order to obtain correct estimations of average nucleaotide frequency which is used for downstream ENC' estimations
$ cat enc.tc.2593339298.con-rib.fna enc.tc.2593339298.con-all.fna   > ../enc.tc.2593339298.con.riball.fna
$ cat enc.tc.2593339298.con-all.fna enc.tc.2593339298.con-all.fna |sed "1s/.*/>con-rib/"  > ../enc.tc.2593339298.con.allall.fna

#remove created intermediate files
$ rm enc*

#calculate ENCprime statistics
$ cd ../
$ ENCprime-master/bin/SeqCount -c enc.tc.2593339298.con.riball.fna 2
$ ENCprime-master/bin/SeqCount -n enc.tc.2593339298.con.allall.fna 2
$ echo -en "\n\n\n0" |ENCprime-master/bin/ENCprime enc.tc.2593339298.con.riball.fna.codcnt enc.tc.2593339298.con.allall.fna.acgtfreq 11 enc.res.2593339298.txt 2
$ cat enc.res.2593339298.txt
Name Nc Ncp ScaledChi SumChi df p B_KM n_codons
con_rib: 32.6589 41.5985 0.4725 3988.8789 43 0.0000 0.5867 8442
con_all: 37.7454 47.7971 0.3040 481132.7170 43 0.0000 0.4525 1582631
Totals: 37.7269 47.7788 0.3043 484187.7488 43 0.0000 0.4526 1591073
```

The ENC prime output together with extracted ribosomal gene coding sequences were used as an input for an downstream R script, to compute &Delta;ENC' (Rocha 2004) and S (Sharp et al. 2005) and return the variables F and d.vs.

```R
library(coRdon) #v1.1.3

# Load sequence data and the ENCprime output statistics
seq.file = list.files(pattern=glob2rx("enc.tc.*.con.riball.fna"))
ENC.file = list.files(pattern=glob2rx("enc.res.*.txt"))

i=1
seq <- readSet(file = seq.file[i])
all <- seq[2]
rib <- seq[1]
seq_mat <- codonCounts(codonTable(all))
rib_mat <- codonCounts(codonTable(rib))

P.phe.a <-sum(seq_mat[, c("TTC")])/sum(seq_mat[, c("TTC", "TTT")]) #proportion of Phe C1 codons in all genes
P.phe.r <-sum(rib_mat[, c("TTC")])/sum(rib_mat[, c("TTC", "TTT")]) #proportion of Phe C1 codons in ribosomal genes

P.ile.a <-sum(seq_mat[, c("ATC")])/sum(seq_mat[, c("ATC", "ATT")]) #proportion of Ile C1 codons in all genes
P.ile.r <-sum(rib_mat[, c("ATC")])/sum(rib_mat[, c("ATC", "ATT")]) #proportion of Ile C1 codons in ribosomal genes

P.tyr.a <-sum(seq_mat[, c("TAC")])/sum(seq_mat[, c("TAC", "TAT")]) #proportion of Tyr C1 codons in all genes
P.tyr.r <-sum(rib_mat[, c("TAC")])/sum(rib_mat[, c("TAC", "TAT")]) #proportion of Tyr C1 codons in ribosomal genes

P.asn.a <-sum(seq_mat[, c("AAC")])/sum(seq_mat[, c("AAC", "AAT")]) #proportion of Asn C1 codons in all genes
P.asn.r <-sum(rib_mat[, c("AAC")])/sum(rib_mat[, c("AAC", "AAT")]) #proportion of Asn C1 codons in ribosomal genes

# Frequency of aminoacid specific codons
freq.phe <- sum(seq_mat[, c("TTC", "TTT")])/sum(seq_mat[, c("TTC", "TTT", "ATC", "ATT", "TAC", "TAT", "AAC", "AAT")])
freq.ile <- sum(seq_mat[, c("ATC", "ATT")])/sum(seq_mat[, c("TTC", "TTT", "ATC", "ATT", "TAC", "TAT", "AAC", "AAT")])
freq.tyr <- sum(seq_mat[, c("TAC", "TAT")])/sum(seq_mat[, c("TTC", "TTT", "ATC", "ATT", "TAC", "TAT", "AAC", "AAT")])
freq.asn <- sum(seq_mat[, c("AAC", "AAT")])/sum(seq_mat[, c("TTC", "TTT", "ATC", "ATT", "TAC", "TAT", "AAC", "AAT")])

# Aminoacid specific S-values
S.phe <-log((P.phe.r*(1-P.phe.a)/P.phe.a)/(1-P.phe.r)) #aminoacid specific S-value
S.ile <-log((P.ile.r*(1-P.ile.a)/P.ile.a)/(1-P.ile.r)) #aminoacid specific S-value
S.tyr <-log((P.tyr.r*(1-P.tyr.a)/P.tyr.a)/(1-P.tyr.r)) #aminoacid specific S-value
S.asn <-log((P.asn.r*(1-P.asn.a)/P.asn.a)/(1-P.asn.r)) #aminoacid specific S-value

# S value, weighted mean of the aminoacid specific S-values
S<- S.phe*freq.phe + S.ile*freq.ile + S.tyr*freq.tyr + S.asn*freq.asn

# Delta ENCprime value
codstat <- read.table (ENC.file[i], sep=' ', header=T)
dENCp <- (codstat[2,3]-codstat[1,3])/codstat[2,3]

# Compute F and predict generation times d.vs as detailed by Vieira-Silva et al. 2010
F <- 6.747*dENCp + 1.184*S -1.438
d.vs <- (1-0.1664*(0.9726-0.7471*(1.184*S+6.747*dENCp-1.438)))^(-1/0.1664)

# GenomeID
IMG.Genome.ID<-substring(seq.file[i],8,17)

# Dataframe and display results
cub.vs=data.frame(IMG.Genome.ID,F,d.vs)
cub.vs
#IMG.Genome.ID         F     d.vs
#2593339298 0.2166112 2.389152
```

Citations:  
* Vieira-Silva, S. and Rocha, E.P.C. (2010) The Systemic Imprint of Growth and Its Uses in Ecological (Meta) Genomics. Plos Genet 6: e1000808.
* Rocha, E.P.C. (2004) Codon usage bias from tRNAâ€™s point of view: Redundancy, specialization, and efficient decoding for translation optimization. Genome Res 14: 2279â€“2286.
* Sharp, P.M., Bailes, E., Grocock, R.J., Peden, J.F., and Sockett, R.E. (2005) Variation in the strength of selected codon usage bias among bacteria. Nucleic Acids Res 33: 1141â€“1153.
* Elek A, Kuzman M, Vlahovicek K (2020). coRdon: Codon Usage Analysis and Prediction of Gene Expressivity. R package version 1.8.0, https://github.com/BioinfoHR/coRdon. 

## Codon Usage bias (gRodon)

Growth rates predicted following a recently published model (Weissman et al 2021) were computed in R using the R package gRodon (https://github.com/jlw-ecoevo/gRodon) (see column 'Generation time (gRodon)' in Table S1). The file 2593339298.genes.fna containing sequences of all predicted genes and the file 2593339298.cog.tab.txt containing functional annotations for identifying genes encoding ribosomal proteins were used as input data.

```R
library(gRodon) #v0.0.0.9000
library(Biostrings) #v2.54.0
setwd('/data/sigmet/gitcode/2593339298/')

# Load sequence and annotation data
seq.file = list.files(pattern=glob2rx("*.genes.fna"))
ann.file = list.files(pattern=glob2rx("*.cog.tab.txt"))

# Predict generation times d.gRodon using the gRodon model
i=1
genes <- readDNAStringSet(seq.file[i])
COG <- read.table(ann.file[i], sep='\t', header=T,quote="")
rib.list <-COG[grep("*Ribosomal protein*", COG$cog_name), 1]
highly_expressed <- gsub('([0-9]+) .*', '\\1', names(genes)) %in% rib.list #edits names by removing everything after first space (including the space)
Gpred <- predictGrowth(genes, highly_expressed)
IMG.Genome.ID<-substring(seq.file[i],1,10)

# Dataframe and display results
cub.gRodon=data.frame(IMG.Genome.ID,d.gRodon=Gpred$d)
cub.gRodon
#IMG.Genome.ID d.gRodon
#2593339298 3.360018
```

Citations:  
* Weissman, J.L., Hou, S., and Fuhrman, J.A. (2021) Estimating maximal microbial growth rates from cultures, metagenomes, and single cells via codon usage patterns. Proc Natl Acad Sci (in print)
* Elek A, Kuzman M, Vlahovicek K (2020). coRdon: Codon Usage Analysis and Prediction of Gene Expressivity. R package version 1.8.0, https://github.com/BioinfoHR/coRdon. 
* PagÃ¨s H, Aboyoun P, Gentleman R, DebRoy S (2020). Biostrings: Efficient manipulation of biological strings. R package version 2.58.0, https://bioconductor.org/packages/Biostrings.
* Henrik Bengtsson (2021). matrixStats: Functions that Apply to Rows and Columns of Matrices (and to Vectors). R package version 0.58.0. https://CRAN.R-project.org/package=matrixStats

## Phylogenetic signal

The phylogenetic signal of each trait was assessed via Pagel's lambda (Pagel 1999) and Blomberg's K (Blomberg et al. 2003) statistics.  We further computed phylogenetic Mantel correlograms similar as detailed elsewhere (Diniz-Filho et al. 2010, Dini-Andreote et al. 2014) to test for significant positive correlations between phylogenetic distances and trait distances at the following phylogenetic distance classes: 0-0.005/0.005-0.01/0.01-0.02/0.02-0.05/0.05-0.1/0.1-0.2/0.2-0.6/0.6-1/1-1.5. 
The traits 'Generation time' and 'Gene duplication' were for all downstream statistics  log(x) transformed and the trait %HGT was log(x+0.01) transformed. Visual inspection of the data  revealed a better fit to normal distribution after these transformation steps. It was not possible to apply transformations for the highly skewed distributions of the number of16s rRNA gene copies or the number of prophages that would have resulted in a better fit to normal distribution. 
The Mantel correlations were tested via the Spearman rank-order correlation as it was not possible to obtain normal distributed trait values in all cases. For the calculation of Mantel correlograms, the dataset was reduced to 10000 randomly selected genomes in order to reduce computation time and memory demand.
For all analyses the phylogeny of the prokaryotic PICRUSt2 default phylogenetic tree (Douglas et a. 2020: pro_ref.tre) was used to infer phylogenetic signals of genomic traits among the PICRUSt2 reference genomes.

The following code returnes the phylogenetic signal of the trait genome size using lambda K statistics as well as the results from Mantel correlograms. These computations were performed analogously for all other traits in Table S1.

```R
library(phytools) #v0.7.20
library(mpmcorrelogram) #v0.1.4
'%!in%' <- function(x,y)!('%in%'(x,y)) #define 'not in' function

# Load data
pic.tre <-read.newick("pro_ref.tre") #Prokaryote reference tree
gtraits <- read.table ("/PATH/TO/TableS1.tab", sep= '\t', header=T) # Trait table (=Table S1)
gtraits <-gtraits[gtraits$Prophages<26 &  gtraits$RRN<20,] #remove outliers: prophages>=26 and RRN>= 20
row.names(gtraits) <- gtraits$PICRUSt.ID

# Select trait for analyses (here: genome size)
trait <-gtraits$Genome.size
names(trait) <-gtraits$PICRUSt.ID

# Test for phylogenetic signal using Blomberg's K statistics
Kstats <-phylosig(pic.tre, na.omit(trait), method="K", test=TRUE, nsim=1001)

# Test for phylogenetic signal using Pagel's lamda statistics
Lstats <-phylosig(pic.tre, na.omit(trait), method="lambda", test=TRUE, nsim=1001)

# Mantel correlogram
n = 10000 # Set n for subsampling
perm = 200 # Set number of permutations for Mantel correlograms
breaks = c(0,0.005,0.01,0.02,0.05,0.1,0.2,0.6,1,1.5) # Set breaks for Mantel correlograms

gtraits <- na.omit(gtraits) # Remove lines with NA from trait table
tips<-sample(pic.tre$tip.label) #all species in tree
tips.n <- sample(gtraits$PICRUST.ID,n) #randomly select 1000- species
tips.notn <- tips[tips %!in% tips.n] #not selected speceis
tre.sub <-fancyTree(pic.tre,type="droptip",tip=tips.notn) #prune tree by removing not selected species
dist.p <- cophenetic(tre.sub) # Compute pairwise phylogenetic distances
dist.t <- dist(trait[names(trait) %in% tre.sub$tip.label], method="manhattan") # Compute pairwise trait value distances
man <- mpmcorrelogram(dist.t, dist.p, method="spearman", permutations=perm, breaks=breaks) # Mantel correlogram
```

Citations:  
* Pagel, M. (1999) Inferring the historical patterns of biological evolution. Nature 401: 877â€“884.
* Blomberg, S.P., Garland, T., and Ives, A.R. (2003) Testing for phylogenetic signal in comparative data: Behavioral traits are more labile. Evolution 57: 717â€“745.
* Dini-Andreote, F., Stegen, J.C., Elsas, J.D. van, and Salles, J.F. (2015) Disentangling mechanisms that mediate the balance between stochastic and deterministic processes in microbial succession. Proc Natl Acad Sci 112: E1326â€“E1332.
* Diniz-Filho, J.A.F., Terribile, L.C., Cruz, M.J.R. da, and Vieira, L.C.G. (2010) Hidden patterns of phylogenetic non-stationarity overwhelm comparative analyses of niche conservatism and divergence. Glob Ecol Biogeogr 19: 916â€“926.
* Douglas, G.M., Maffei, V.J., Zaneveld, J.R., Yurgel, S.N., Brown, J.R., Taylor, C.M., et al. (2020) PICRUSt2 for prediction of metagenome functions. Nat Biotechnol 38: 685â€“688.
